better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1129 @@
+- in `mutating` methods and in the mutation clauses of subscripts
+  and computed properties, ensure the invariant is intact:
+  - before returning to the caller
+  - before passing `self` as a parameter
+  - before calling a non-private method or subscript—or accessing a non-private computed property—of `self`.
+
+Non-mutating operations can't disturb the invariant, and any mutating
+operation that only uses the type's public API trivially upholds the
+invariant.
+
+#### How To Choose a Type Invariant
+
+Often, you have a choice about how strong to make your type's
+invariant.  For example, in `PairArray`, we could have chosen the
+invariant `x.count <= y.s.count`.  We call that a *weaker* invariant
+because it allows many more internal states.
+
+Let's assume that there's some method that can create a condition
+where `x.count < y.count`:
+
+```swift
+  /// An instance with the value of `zip(xs, ys)`.
+  ///
+  /// - Precondition: `xs.count <= ys.count`.
+  init(xs: [X], ys: [Y]) { (self.xs, self.ys) = (xs, ys) }
+```
+
+[Note: when sequences of unequal length are `zip`ped, the result
+has the length of the shorter sequence.]
+
+Now let's see what happens to the rest of the implementation.  Most of
+it is unchanged, but in the `append` method, we now need to account
+for these new internal states.  Here's one way we could do it:
+
+```swift
+  /// Adds `e` to the end.
+  public mutating func append(_ e: (X, Y)) {
+    ys.removeRange(xs.count...) // <=====
+    xs.append(e.0)
+    ys.append(e.1)
+  }
+```
+
+Also, the `count` property now reports the wrong length; it needs to
+return `xs.count` instead of `ys.count`.
+
+So a weaker invariant complicated the implementation and made it more
+fragile. In this case, only two adjustments were needed, but in
+principle it could be many more. Maintaining a stronger invariant
+is better.  There are two simple approaches:
+
+- We can strengthen the precondition of the new `init` method:
+
+  ```swift
+    /// An instance with the value of `zip(xs, ys)`.
+    ///
+    /// - Precondition: `xs.count == ys.count`.
+    init(xs: [X], ys: [Y]) { (self.xs, self.ys) = (xs, ys) }
+  ```
+
+- Or we can lift the precondition entirely and normalize the internal
+  representation upon construction:
+
+  ```swift
+    /// An instance with the value of `zip(xs, ys)`.
+    init(xs: [X], ys: [Y]) {
+      (self.xs, self.ys) = (xs, ys)
+      let l = Swift.min(xs.count, ys.count)
+      self.xs.removeRange(l...)
+      self.ys.removeRange(l...)
+    }
+  ```
+
+The latter approach is better, as it results in a simpler and
+less-fragile public API.
+
+### It's Documentation
+
+Looking back at Meyers' definition, you might notice he says contracts
+are “precisely defined specifications,” which is just a fancy word for
+documentation. Like it or not, if you're going to be in the
+correctness game, you need documentation.  Undocumented software is
+neither correct nor incorrect; it does something, but does it do the
```

Comment:
I think this is too strong a claim (and too fundamental to the BC philosophy) to make in such a cheeky (for lack of a better term) tone. "Like it or not" seems to take an adversarial stance - assuming the reader is anti-documentation, which is a bit off-putting. And even if they are, I worry this tone would make them subconsciously dig in their heels. 

I also think that this deserves an example - perhaps even a callback to the decision space around the `PairArray` API - so we can show the reader firsthand that evaluating the correctness of an undocumented component is an impossible task. 

I appreciate this is general feedback without a concrete suggestion - let me know if you want to workshop this.

Reply:
I don't know what claim you're saying is too strong, or what you think I should say instead.

I agree with your criticism of the tone.  As for examples, I don't understand how anyone can demonstrate that evaluating correctness is impossible without documentation.  Where would you even start in trying to evaluate correctness?  If we write the example without docs that would be the first thing we'd confront.